### ISSUE TITLE
`suspicious-doc-comment` shouldn't trigger for `#[doc = "!.."]` comments

### ISSUE THREAD

**Issue by JarredAllen** [rust-lang/rust-clippy#14265]
### Summary

The `suspicious-doc-comments` lint triggers on comments written as `#[doc = "!.."]` (where `"!.."` is a string literal beginning with an `!` character), thinking that you meant to write `//!..` as an inner doc comment. This attribute isn't meant as an inner doc comment, since the `!` is farther removed from where it would change the comment type, compared to the `//!`/`///` form (where this lint makes sense).

While this isn't something that someone would likely write directly, I have macros where the doc comments include running `stringify!` or `concat!` on the macro input, which can result in a doc comment that starts with an `!`, and these comments now raise lints with the 1.85 release.

### Lint Name

suspicious-doc-comments

### Reproducer

I tried this code ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2bdd448fa9efff6461237c4ef9a9cc1e)):

```rust
#[doc = "! here's some docs !"]
fn main() {}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: this is an outer doc comment and does not apply to the parent module or crate
 --> src/main.rs:1:1
  |
1 | #[doc = "! here's some docs !"]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#suspicious_doc_comments
  = note: `#[warn(clippy::suspicious_doc_comments)]` on by default
help: use an inner doc comment to document the parent module or crate
  |
1 | //! here's some docs !
  |
```

I expected to see this happen: No lints

### Version

```text
rustc 1.85.0 (4d91de4e4 2025-02-17)
binary: rustc
commit-hash: 4d91de4e48198da2e33413efdcd9cd2cc0c46688
commit-date: 2025-02-17
host: x86_64-unknown-linux-gnu
release: 1.85.0
LLVM version: 19.1.7

(this was taken on my machine, but I also see the behavior on the current playground)
```

**Comment by samueltardieu** [comment]
This lint is here to intercept uses of

```rust
///! here's some docs !
fn main() {}
```

which is transformed by early passes of the compiler into the code you show.

Unless this pattern is common because it comes from a popular macro-expanding crate, maybe you should add a `#![allow(clippy::suspicious_doc_comment)]` to your crate, or  a `suspicious-doc-comment = allow` in the `[lints.clippy]` section if your `Cargo.toml`.

**Comment by samueltardieu** [comment]
Fixed by the latest rustup, nightly Clippy versions will get the fix.
